### PR TITLE
#36 - Save locations for easy access 

### PR DESCRIPTION
**Pull request by meshgin**
A user may save/unsave their favorite places upon searching for a location and clicking on the save button. 
They may then view an updated list of of their saved places by going to My Places from the nav menu.

### CODE REVIEW

File: app/src/main/java/com/droidhats/campuscompass/adapters/FavoritesAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import com.google.android.gms.maps.model.LatLng
+
+class FavoritesAdapter(
+    private val items: List<FavoritePlace>,
+    private val listener: OnFavoriteClickListener?,
+    private val fragment: MyPlacesFragment?
+) : RecyclerView.Adapter<FavoritesAdapter.ViewHolder>() {
+
+    private var onClickListener: View.OnClickListener
+    private lateinit var viewModel: FavoritesViewModel
```
Comment (Neeham) [inline app/src/main/java/com/droidhats/campuscompass/adapters/FavoritesAdapter.kt:25]: You can safely remove this as it is not being used 😄

File: app/src/main/java/com/droidhats/campuscompass/adapters/FavoritesAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        }
+    }
+
+    override fun getItemCount(): Int = items.size
+
+    inner class ViewHolder(val view: View) : RecyclerView.ViewHolder(view) {
+
+        val titleView: TextView = view.findViewById(R.id.favorites_title_item)
+        val locationView: TextView = view.findViewById(R.id.favorites_location_item)
+        var cardView: CardView = view.findViewById(R.id.favorites_card_view)
```
Comment (Neeham) [inline app/src/main/java/com/droidhats/campuscompass/adapters/FavoritesAdapter.kt:71]: You can safely remove this as it is not being used 😄

File: app/src/main/java/com/droidhats/campuscompass/repositories/FavoritesRepository.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        if (repository.containsKey(placeID)) {
+            return repository[placeID]
+        } else {
+            return null
+        }
+    }
+
+    fun listAllPlaces() : List<FavoritePlace> = repository.values.toList()
+
+    fun createFavoritePlace(googlePlace: GooglePlace) : FavoritePlace {
```
Comment (Neeham) [inline app/src/main/java/com/droidhats/campuscompass/repositories/FavoritesRepository.kt:39]: You can make this function a Private function 😄

File: app/src/main/java/com/droidhats/campuscompass/viewmodels/SearchViewModel.kt [hunk 4]
```diff
@@ -194,11 +194,15 @@ class SearchViewModel(application: Application) : AndroidViewModel(application)
      * The distance returned is in km
      */
     private fun haversine(location1 : Location, location2 : Location) : Double{
-        val diffLat = Math.toRadians(location2.getLocation().latitude - location1.getLocation().latitude)
-        val diffLong = Math.toRadians(location2.getLocation().longitude- location1.getLocation().longitude)
+        return haversine(location1.getLocation(), location2.getLocation())
+    }
+
+    fun haversine(latLng1: LatLng, latLng2: LatLng) : Double {
```
Comment (Neeham) [inline app/src/main/java/com/droidhats/campuscompass/viewmodels/SearchViewModel.kt:200]: You can make this function a Private function 😄

File: app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt [hunk 5]
```diff
@@ -591,15 +599,42 @@ class MapFragment : Fragment(), OnMapReadyCallback, GoogleMap.OnMarkerClickListe
         map!!.animateCamera(CameraUpdateFactory.newLatLngZoom(coordinates, zoomLevel))
     }
 
+    private fun toggleSaveButton() {
+        val saveButton : Button = requireActivity().findViewById(R.id.place_card_favorites_button)
+        if (saveButton.text.equals("Save")) {
```
Comment (Neeham) [inline app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt:604]: Please change this to binary operator:

`if (saveButton.text == "Save") {`

File: app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt [hunk 6]
```diff
@@ -653,6 +700,22 @@ class MapFragment : Fragment(), OnMapReadyCallback, GoogleMap.OnMarkerClickListe
             observer?.update(map!!.cameraPosition.zoom)
         }
     }
+
+    override fun onFavoriteClick(item: FavoritePlace?) {
+        if (item != null) {
+            var googlePlace : GooglePlace = GooglePlace(item.placeId, item.name?:"", item.address?:"", LatLng(item.latitude, item.longitude))
```
Comment (Neeham) [inline app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt:706]: Please declare this immutable changing `var `to `val`

File: app/src/main/java/com/droidhats/campuscompass/views/MyPlacesFragment.kt [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    private fun updateRecyclerView() {
+        val places : List<FavoritePlace> = favoritesViewModel.getFavorites()
+        val fragment : MyPlacesFragment = this
+        with(recyclerView) {
+            layoutManager = LinearLayoutManager(context)
+            adapter = FavoritesAdapter(places, onFavoriteClickListener, fragment)
+        }
+    }
+
+    fun getCurrentLocation() {
```
Comment (Neeham) [inline app/src/main/java/com/droidhats/campuscompass/views/MyPlacesFragment.kt:69]: Please make this a private function 😄

File: app/src/main/res/layout/my_places_fragment.xml [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+            app:layoutManager="LinearLayoutManager"
+            tools:context=".views.MyPlacesFragment"
+            tools:listitem="@layout/favorites_recycler_item" />
+
+    <LinearLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:orientation="vertical"
+        android:paddingBottom="15dp"
+        android:paddingTop="15dp"/>
```
Comment (Neeham) [inline app/src/main/res/layout/my_places_fragment.xml:62]: Can you please remove this? This is not needed at all.

File: app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt [hunk 9]
```diff
@@ -591,15 +599,42 @@ class MapFragment : Fragment(), OnMapReadyCallback, GoogleMap.OnMarkerClickListe
         map!!.animateCamera(CameraUpdateFactory.newLatLngZoom(coordinates, zoomLevel))
     }
 
+    private fun toggleSaveButton() {
+        val saveButton : Button = requireActivity().findViewById(R.id.place_card_favorites_button)
+        if (saveButton.text.equals("Save")) {
+            saveButton.text = "Saved"
+            val redHeart : Drawable? = context?.resources?.getDrawable(R.drawable.ic_favorites_heart_red)
```
Comment (Neeham) [inline app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt:606]: Avoid using depreciated method as much as possible 😄 

Please replace: 

`val redHeart : Drawable? = context?.resources?.getDrawable(R.drawable.ic_favorites_heart_red)`

with:

`val redHeart = activity?.let { ContextCompat.getDrawable(it, R.drawable.ic_favorites_heart_red) }`

File: app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt [hunk 10]
```diff
@@ -591,15 +599,42 @@ class MapFragment : Fragment(), OnMapReadyCallback, GoogleMap.OnMarkerClickListe
...
+            saveButton.setCompoundDrawablesWithIntrinsicBounds(redHeart, null, null, null)
+        } else {
+            saveButton.text = "Save"
+            val greyHeart : Drawable? = context?.resources?.getDrawable(R.drawable.ic_favorites_heart_grey)
```
Comment (Neeham) [inline app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt:610]: Avoid using depreciated method as much as possible 😄 

Please replace: 

`val greyHeart : Drawable? = context?.resources?.getDrawable(R.drawable.ic_favorites_heart_grey)`

with:

`val greyHeart = activity?.let { ContextCompat.getDrawable(it, R.drawable.ic_favorites_heart_grey) }`

File: app/src/main/java/com/droidhats/campuscompass/repositories/FavoritesRepository.kt [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        val name : String = googlePlace.name
+        val latitude : Double = googlePlace.coordinate.latitude
+        val longitude : Double =  googlePlace.coordinate.longitude
+        val address : String? = googlePlace.place?.address
+        val placeId : String = googlePlace.placeID
+
+        return FavoritePlace(placeId, name, latitude, longitude, address)
+    }
+}
```
Comment (Neeham) [inline app/src/main/java/com/droidhats/campuscompass/repositories/FavoritesRepository.kt:48]: I don't think this class file is being used at all in our code. Could you please double verify to see if we really need this? 😄

File: app/src/main/java/com/droidhats/campuscompass/adapters/FavoritesAdapter.kt [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+            // Notify the activity/fragment that an item has been clicked
+             listener?.onFavoriteClick(item)
+        }
+
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        val view = LayoutInflater.from(parent.context).inflate(R.layout.favorites_recycler_item, parent, false)
+
+
```
Comment (JeremiahTee) [inline app/src/main/java/com/droidhats/campuscompass/adapters/FavoritesAdapter.kt:39]: Could you remove this line for formatting purposes? :)

File: app/src/main/java/com/droidhats/campuscompass/roomdb/FavoritesDatabase.kt [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+import androidx.room.RoomDatabase
+import com.droidhats.campuscompass.models.FavoritePlace
+import com.droidhats.campuscompass.repositories.FavoritePlacesDao
+
+@Database(entities = [FavoritePlace::class], version = 2, exportSchema = true)
+abstract class FavoritesDatabase : RoomDatabase() {
+    abstract fun favoritePlacesDao(): FavoritePlacesDao
+
+    companion object {
+
```
Comment (JeremiahTee) [inline app/src/main/java/com/droidhats/campuscompass/roomdb/FavoritesDatabase.kt:15]: If possible remove this extra empty line for format consistency

File: app/src/main/java/com/droidhats/campuscompass/viewmodels/FavoritesViewModel.kt [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+import com.google.android.gms.maps.model.LatLng
+import kotlin.math.sin
+import kotlin.math.sqrt
+import kotlin.math.cos
+import kotlin.math.asin
+import kotlin.math.pow
+
+/*
+  This class must extend AndroidViewModel instead of just ViewModel because
+  ContentResolver requires the application context to be able to query calendar info
```
Comment (JeremiahTee) [inline app/src/main/java/com/droidhats/campuscompass/viewmodels/FavoritesViewModel.kt:16]: Update this line comment with "ContentResolver requires the application context to be able to query **favorite place info**"

File: app/src/main/java/com/droidhats/campuscompass/repositories/FavoritesRepository.kt (hunk 11 above)
Comment (JeremiahTee) [reply app/src/main/java/com/droidhats/campuscompass/repositories/FavoritesRepository.kt:48]: If I'm not mistaken this class will be used for the unit test in the same fashion as CalendarRepository.kt and CalendaryRepositoryTest.kt, could confirm with @makchamp if the unit test is needed

File: app/src/main/java/com/droidhats/campuscompass/viewmodels/MapViewModel.kt [hunk 15]
```diff
@@ -28,14 +29,19 @@ class  MapViewModel(application: Application) : AndroidViewModel(application) {
...
     private var buildings: List<Building>
+    private val favoritesDb : FavoritesDatabase
 
     init {
         mapRepository = MapRepository.getInstance(context)
         navigationRepository = NavigationRepository.getInstance(getApplication())
         campuses = mapRepository.getCampuses()
         buildings = mapRepository.getBuildings()
+        favoritesDb = FavoritesDatabase.getInstance(context)
+
```
Comment (JeremiahTee) [inline app/src/main/java/com/droidhats/campuscompass/viewmodels/MapViewModel.kt:40]: Remove this line please :)

File: app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt [hunk 16]
```diff
@@ -591,15 +599,42 @@ class MapFragment : Fragment(), OnMapReadyCallback, GoogleMap.OnMarkerClickListe
...
+            saveButton.setCompoundDrawablesWithIntrinsicBounds(greyHeart, null, null, null)
+        }
+
```
Comment (JeremiahTee) [inline app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt:613]: Would remove this empty line as well

File: app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt [hunk 17]
```diff
@@ -591,15 +599,42 @@ class MapFragment : Fragment(), OnMapReadyCallback, GoogleMap.OnMarkerClickListe
...
+        val foundPlace : FavoritePlace? = favoritesDb.favoritePlacesDao().findPlaceById(location.placeID)
+        if (foundPlace != null) {
+            saveButton.text = "Saved"
+            val redHeart : Drawable? = context?.resources?.getDrawable(R.drawable.ic_favorites_heart_red)
+            saveButton.setCompoundDrawablesWithIntrinsicBounds(redHeart, null, null, null)
+        } else {
+            saveButton.text = "Save"
+            val greyHeart : Drawable? = context?.resources?.getDrawable(R.drawable.ic_favorites_heart_grey)
+            saveButton.setCompoundDrawablesWithIntrinsicBounds(greyHeart, null, null, null)
+
```
Comment (JeremiahTee) [inline app/src/main/java/com/droidhats/campuscompass/views/MapFragment.kt:635]: Would remove this empty line as well

File: app/src/main/java/com/droidhats/campuscompass/views/MyPlacesFragment.kt [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+        var location : LatLng? = null
+        val fusedLocationClient = LocationServices.getFusedLocationProviderClient(activity as Activity)
+        fusedLocationClient.lastLocation.addOnSuccessListener{
+            location = LatLng(it.latitude, it.longitude)
+        }
+
+        if (location == null) {
```
Comment (JeremiahTee) [inline app/src/main/java/com/droidhats/campuscompass/views/MyPlacesFragment.kt:76]: As discussed in the daily meeting this morning, this was temporarily put because the current location received is always null. After reviewing the code, I concluded that it's because of the way the Task is being handled in Line 71:
` val fusedLocationClient = LocationServices.getFusedLocationProviderClient(activity as Activity)
        fusedLocationClient.lastLocation.addOnSuccessListener{
            location = LatLng(it.latitude, it.longitude)
        }`

If I'm not mistaken @makchamp has experience with Tasks so I will discuss with him on how to fix this

**Comment by JeremiahTee** [comment]
Overall great changes! I like this feature, it's really handy. I only have a few comments

_Functionality-wise:_
I noticed that when you press on the directions button on a saved place, it brings you to the search fragment (as intended) but the destination is an address & the navigation button cannot be clicked (it is not highlighted - Figure 1). You have to press again on the address listed below to enable the navigation toggle. Would it be possible to look into this and fix it? 
I compared with when pressing directions on a place from the info card (Figure 2), it brings you to search fragment and the name of the place is displayed (instead of the address) and the navigation button is highlighted (can be clicked) -> Figure 3.

**Figure 1**
Fig1

**Figure 2**
Fig2

**Figure 3**
Fig3

TLDR;
Figure 1 should be like Figure 3 -> the navigation button is highlighted upon pressing directions to a place & the destination textbox should display the name of the place instead of the address (debatable)

_Code-wise:_
Some formatting nitpicks. Mostly empty lines. 
As discussed in the meeting, a fix will be discussed to "fetch the current location" - as I wrote in my last inline comment, it has to do with handling a Task.

File: app/src/main/java/com/droidhats/campuscompass/views/MyPlacesFragment.kt (hunk 18 above)
Comment (JeremiahTee) [reply app/src/main/java/com/droidhats/campuscompass/views/MyPlacesFragment.kt]: Fixed with 
f332077

**Comment by JeremiahTee** [comment]
> _Functionality-wise:_
> I noticed that when you press on the directions button on a saved place, it brings you to the search fragment (as intended) but the destination is an address & the navigation button cannot be clicked (it is not highlighted - Figure 1). 

Fixed with ff83447 during a pair programming session. Thank you for the fix @makchamp 

Only minor formatting comments are left  😄

File: app/src/main/java/com/droidhats/campuscompass/views/SearchFragment.kt [hunk 19]
```diff
@@ -24,11 +24,7 @@ import androidx.recyclerview.widget.LinearLayoutManager
 import androidx.recyclerview.widget.RecyclerView
 import com.droidhats.campuscompass.R
 import com.droidhats.campuscompass.adapters.SearchAdapter
-import com.droidhats.campuscompass.models.Location
-import com.droidhats.campuscompass.models.GooglePlace
-import com.droidhats.campuscompass.models.Building
-import com.droidhats.campuscompass.models.IndoorLocation
-import com.droidhats.campuscompass.models.NavigationRoute
+import com.droidhats.campuscompass.models.*
```
Comment (JeremiahTee) [inline app/src/main/java/com/droidhats/campuscompass/views/SearchFragment.kt:27]: Oops these imports got nuked somehow by the IDE in one of my commits. Please fix this :)

File: app/src/main/java/com/droidhats/campuscompass/views/SearchFragment.kt [hunk 20]
```diff
@@ -113,6 +109,20 @@ class SearchFragment : Fragment()  {
...
+                GooglePlace(
+                    favPlaceArg.placeId,
+                    favPlaceArg.name,
+                    "",
+                    LatLng(favPlaceArg.latitude, favPlaceArg.longitude)
+                ), true
+            )
+            arguments?.clear()
+        }
+
```
Comment (JeremiahTee) [inline app/src/main/java/com/droidhats/campuscompass/views/SearchFragment.kt:125]: Forgot to remove this extra empty line, this is the last one I promise 😅

File: app/src/main/java/com/droidhats/campuscompass/repositories/FavoritesRepository.kt (hunk 11 above)
Comment (meshgin) [reply app/src/main/java/com/droidhats/campuscompass/repositories/FavoritesRepository.kt:48]: I could remove the code here because initially i was thinking of using FavoritesRepository to implement the data storage but instead i used Room and the operations described in FavoritePlacesDao.

**Comment by JeremiahTee** [comment]
When pressing the navigation button from a place card in the My Place fragment, the destination textbox shows the name of the place AND its respectful address. Thanks @makchamp for the fix -> 
7604a6f

All the comments have been addressed by @meshgin . I fixed the conflicts with the master branch and committed a last refactoring change. Will merge now 🛩️
